lib/ecto/repo/schema.ex

```diff
@@ -591,7 +591,16 @@ defmodule Ecto.Repo.Schema do
       {:invalid, _} = constraints ->
         constraints
       {:error, :stale} ->
-        raise Ecto.StaleEntryError, struct: changeset.data, action: action
+        opts = List.last(args)
+
+        case Keyword.fetch(opts, :stale_error_field) do
+          {:ok, stale_error_field} ->
+            changeset = Ecto.Changeset.add_error(changeset, stale_error_field, "stale")
```

Comment:
great work @qcam. I'd like to propose a couple minor changes:

1. we should probably change clause to be a bit stricter:

   ```elixir
   {:ok, stale_error_field} when is_atom(stale_error_field) ->
   ```

2. add error metadata, e.g.: `add_error(changeset, stale_error_field, message, [stale: true])`. This way, it's similar to `:validation` and `:constraint` error metadata keys, so we can programatically handle this error without trying to look for the message string.

3. change `"stale"` to `"is stale"`, or maybe even: `Keyword.get(opts, :stale_error_message, "is stale")`

wdyt?

Reply:
@wojtekmach your proposal sounds great, I'm up for changing but the PR is merged. Shall I open a new one with these changes?

Reply:
Yes, please open a new one! :D